modules/cache/cache.go

```diff
@@ -35,6 +36,20 @@ func NewContext() error {
 		if conn, err = newCache(setting.CacheService.Cache); err != nil {
 			return err
 		}
+		const testKey = "__gitea_cache_test"
+		const testVal = "test-value"
+		if err = conn.Put(testKey, testVal, 10); err != nil {
+			return err
+		}
+		val := conn.Get(testKey)
+		if valStr, ok := val.(string); !ok || valStr != testVal {
+			// If the cache is full, the Get may not read the expected value stored by Put.
+			// Since we have checked that Put can success, so we just show a warning here, do not return an error to panic.
+			log.Warn("cache (adapter:%s, config:%s) doesn't seem to work correctly, set test value '%v' but get '%v'",
```

Comment:
The **Get** method in [file](https://gitea.com/go-chi/cache/src/branch/master/file.go#L87-L98), [memcache](https://gitea.com/go-chi/cache/src/branch/master/memcache/memcache.go#L47-L53), [redis](https://gitea.com/go-chi/cache/src/branch/master/redis/redis.go#L63-L69), [mysql](https://gitea.com/go-chi/cache/src/branch/master/mysql/mysql.go#L84-L96), [postgres](https://gitea.com/go-chi/cache/src/branch/master/postgres/postgres.go#L84-L96) either returns the value or **nil**. It will return **nil** if the key is not found in the cache or if there is an error (including a connection error) and there is no way to distinguish between the two.

As @zeripath pointed out, returning **nil** does not mean there is a malfunction and issuing a warning could be misleading to the system administrator. If their cache is very busy, they may see this warning intermittently even though all is working as expected.

Maybe it is best to not test with **Get** at all? Or to issue a log.Info message declaring that the cache has been verified to work when it returns the expected value and remain silent when the return value is nil?

Reply:
If the cache is busy, if the value just written in can not be read, I believe the cache doesn't work correctly.

In such case, the sys admin should tune the cache system ASAP, otherwise, the cache doesn't help at all.

Reply:
In that case it may be worth explaining more precisely. As a system administrator I would not know what to make of the fact that storing ABC returns nil. Maybe something like **" adding key=%s value=%s in the cache worked but the getting the value of key=%s immediately afterwards failed. It is likely because the cache is too busy or not working properly.", testKey, testVal, testKey** ?

Reply:
In my mind the log should only show the problem, but not guess the reasons for sys admins. Incorrect guess would mislead users. A sys admin should be experienced and know how to check the caches. If the sys admins don't know how to check the cache, then they can ignore this warning, Gitea could still work.

Reply:
> In my mind the log should only show the problem, but not guess the reasons for sys admins.

Fair enough.

>  ...then they can ignore this warning, Gitea could still work.

The error message does not say that though: how can the sysadmin know they can ignore the warning?

Reply:
Well, it's just a warning, not a error, a warning usually means that it's not that serious .... and Gitea continues running.

Reply:
I'm sorry, it looks like I've wasted your time trying to improve the readability of this error message.

Reply:
Oh you are welcome, I think the suggestions are good. Open discussion is always good.